doc/manual/src/contributing/hacking.md

```diff
@@ -8,50 +8,82 @@ $ git clone https://github.com/NixOS/nix.git
 $ cd nix
 ```
 
-To build Nix for the current operating system/architecture use
+## Nix with flakes
+
+This section assumes you are using Nix with flakes enabled. See the next section for equivalent instructions which don't require flakes.
+
+
+To build all dependencies and start a shell in which all environment
+variables are set up so that those dependencies can be found:
 
 ```console
-$ nix-build
+$ nix develop
 ```
 
-or if you have a flake-enabled nix:
+This shell also adds `./outputs/bin/nix` to your `$PATH` so you can run `nix` immediately after building it.
+
+To get a shell with a different compilation environment (e.g. `stdenv`,
+`gccStdenv`, `clangStdenv`, `clang11Stdenv`, `ccacheStdenv`):
 
 ```console
-$ nix build
+$ nix develop .#clang11Stdenv
 ```
 
-This will build `defaultPackage` attribute defined in the `flake.nix`
-file. To build for other platforms add one of the following suffixes to
-it: aarch64-linux, i686-linux, x86\_64-darwin, x86\_64-linux. i.e.
+Note: you can use `ccacheStdenv` to drastically improve rebuild
+time. By default, ccache keeps artifacts in `~/.cache/ccache/`.
+
+To build Nix itself in this shell:
 
 ```console
-$ nix-build -A defaultPackage.x86_64-linux
+[nix-shell]$ ./bootstrap.sh
+[nix-shell]$ ./configure $configureFlags --prefix=$(pwd)/outputs/out
+[nix-shell]$ make -j $NIX_BUILD_CORES
 ```
 
-To build all dependencies and start a shell in which all environment
-variables are set up so that those dependencies can be found:
+To install it in `$(pwd)/outputs` and test it:
 
 ```console
-$ nix-shell
+[nix-shell]$ make install
+[nix-shell]$ make installcheck -j $NIX_BUILD_CORES
+[nix-shell]$ nix --version
+nix (Nix) 2.12
 ```
 
-or if you have a flake-enabled nix:
+To build a release version of Nix, use:
 
 ```console
-$ nix develop
+$ nix build
 ```
 
-To get a shell with a different compilation environment (e.g. stdenv,
-gccStdenv, clangStdenv, clang11Stdenv, ccacheStdenv):
+This will build the `packages.${builtins.currentSystem}.default` attribute defined in the `flake.nix` file and make the result available under `./result/bin`.
+
+To build for other platforms, use one of the following platform identifiers:
+
+- `aarch64-linux`
+- `i686-linux`
+- `x86_64-darwin`
+- `x86_64-linux`
```

Comment:
```suggestion
To build for [other platforms](#target-platforms), specify a platform identifier:
```

Please add another section below with the platform list and the little explanation that without parameters Nix will build for the current platform. Also change that in the other section for "classic Nix".